python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -181,4 +201,132 @@ def __init__(
+            and scaling_config.num_gpus_per_worker <= 0
+        ):
+            logger.info(
+                "GPUs are detected in your Ray cluster, but GPU "
+                "training is not enabled for Ray Train. To enable "
+                "GPU training, make sure to set `use_gpu` to True "
+                "when instantiating your Trainer."
+            )
+
+        self.train_loop_per_worker = train_loop_per_worker
+        self.train_loop_config = train_loop_config
+
+        super(DataParallelTrainer, self).__init__(
+            scaling_config=scaling_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+        if "num_workers" not in self.scaling_config:
+            raise ValueError("You must specify the 'num_workers' in scaling_config.")
+
+        if self.scaling_config["num_workers"] < 0:
+            raise ValueError("'num_workers' must be a non-negative integer.")
+
+        num_params = len(inspect.signature(self.train_loop_per_worker).parameters)
+        if num_params > 1:
+            raise ValueError(
+                f"train_loop_per_worker should take in 0 or 1 arguments, "
+                f"but it accepts {num_params} arguments instead."
+            )
+
+        backend_config = backend_config if backend_config else BackendConfig()
+        self.backend_config = backend_config
+
+        self.train_env_var_values = {
+            var_name: os.environ[var_name]
+            for var_name in _BACKEND_ENV_VARS
+            if var_name in os.environ
+        }
+
+    def training_loop(self) -> None:
+        scaling_config_dataclass = ScalingConfigDataClass(**self.scaling_config)
+
+        train_loop_per_worker = construct_train_func(
+            self.train_loop_per_worker,
+            self.train_loop_config,
+            fn_arg_name="train_loop_per_worker",
+        )
+
+        runtime_env = {"env_vars": self.train_env_var_values}
+
+        remote_executor = ray.remote(num_cpus=0)(BackendExecutor)
+        additional_resources_per_worker = (
+            scaling_config_dataclass.additional_resources_per_worker
+        )
+
+        backend_executor_actor = remote_executor.options(
+            runtime_env=runtime_env
+        ).remote(
+            backend_config=self.backend_config,
+            num_workers=scaling_config_dataclass.num_workers,
+            num_cpus_per_worker=scaling_config_dataclass.num_cpus_per_worker,
+            num_gpus_per_worker=scaling_config_dataclass.num_gpus_per_worker,
+            additional_resources_per_worker=additional_resources_per_worker,
+            max_retries=0,
+        )
+
+        checkpoint_manager = MLTuneCheckpointManager()
+        checkpoint_manager.on_init(preprocessor=self.preprocessor)
+
+        # Start the remote actors.
+        ray.get(backend_executor_actor.start.remote(initialization_hook=None))
```

Comment:
Just for my understanding, is this launching a separate actor? So there is the Trainer actor, the BackendExecutor actor, and also the worker actors?

Reply:
What if we ran the backend executor directly in this worker?

Reply:
Yes we definitely should run the backend executor directly in the Trainable! This does require some non-trivial refactoring of Ray Train though, so let me add a TODO for this and I will follow up in a subsequent PR.